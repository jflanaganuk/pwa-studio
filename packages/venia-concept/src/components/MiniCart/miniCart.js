import React from 'react';
import { bool, func, shape, string } from 'prop-types';

import { useWindowSize } from '@magento/peregrine';

import Body from './body';
import Footer from './footer';
import Header from './header';
import Mask from './mask';

import { mergeClasses } from 'src/classify';
import defaultClasses from './miniCart.css';

const MiniCart = props => {
    // Props.
    const {
        beginEditItem,
        cancelCheckout,
        cart,
        closeDrawer,
        endEditItem,
        isCartEmpty,
        isMiniCartMaskOpen,
        isOpen,
        removeItemFromCart,
        updateItemInCart
    } = props;
    const { isEditingItem, isLoading } = cart;

    // Hooks.
    const { innerHeight: viewportHeight } = useWindowSize();

    // Members.
    const classes = mergeClasses(defaultClasses, props.classes);
    const rootClass = isOpen ? classes.root_open : classes.root;
    const rootStyle = {
        '--minicart-height-unit': `${viewportHeight * 0.01}px`
    };

    const showFooter = !(isCartEmpty || isLoading || isEditingItem);
    const footer = showFooter ? (
        <Footer cart={cart} isMiniCartMaskOpen={isMiniCartMaskOpen} />
    ) : null;

    return (
        <aside className={rootClass} style={rootStyle}>
            <Header closeDrawer={closeDrawer} isEditingItem={isEditingItem} />
            <Body
                beginEditItem={beginEditItem}
                cart={cart}
                closeDrawer={closeDrawer}
                endEditItem={endEditItem}
                isCartEmpty={isCartEmpty}
                removeItemFromCart={removeItemFromCart}
                updateItemInCart={updateItemInCart}
            />
            <Mask isActive={isMiniCartMaskOpen} dismiss={cancelCheckout} />
            {footer}
        </aside>
    );
};

MiniCart.propTypes = {
    beginEditItem: func.isRequired,
    cancelCheckout: func,
    cart: shape({
        isEditingItem: bool,
        isLoading: bool
    }).isRequired,
    classes: shape({
        header: string,
        root: string,
        root_open: string,
        title: string
    }),
    closeDrawer: func,
    endEditItem: func.isRequired,
    isCartEmpty: bool,
    isMiniCartMaskOpen: bool,
    isOpen: bool,
    removeItemFromCart: func,
    updateItemInCart: func
};

export default MiniCart;
