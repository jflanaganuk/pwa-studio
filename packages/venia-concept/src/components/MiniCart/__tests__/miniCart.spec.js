import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { useWindowSize } from '@magento/peregrine';

import MiniCart from '../miniCart';

const renderer = new ShallowRenderer();

test('renders the correct tree', () => {
    const props = {
        beginEditItem: jest.fn(),
        endEditItem: jest.fn(),
        cart: {
            isEditingItem: false
        }
    };

    useWindowSize.mockReturnValueOnce({ innerHeight: 719 });

    const tree = renderer.render(<MiniCart {...props} />);

    expect(tree).toMatchSnapshot();
});

test('doesnt render a footer when cart is empty', () => {
    const props = {
        beginEditItem: jest.fn(),
        cart: {
            isEditingItem: false
        },
        endEditItem: jest.fn(),
        isCartEmpty: true
    };

    renderer.render(<MiniCart {...props} />);
    const result = renderer.getRenderOutput();

    const footer = result.props.children[3];
    expect(footer).toBeNull();
});

test('doesnt render a footer when cart is editing', () => {
    const props = {
        beginEditItem: jest.fn(),
        cart: {
            isEditingItem: true
        },
        endEditItem: jest.fn()
    };

    renderer.render(<MiniCart {...props} />);
    const result = renderer.getRenderOutput();

    const footer = result.props.children[3];
    expect(footer).toBeNull();
});

test('doesnt render a footer when cart is loading', () => {
    const props = {
        beginEditItem: jest.fn(),
        cart: {
            isLoading: true
        },
        endEditItem: jest.fn()
    };

    renderer.render(<MiniCart {...props} />);
    const result = renderer.getRenderOutput();

    const footer = result.props.children[3];
    expect(footer).toBeNull();
});
