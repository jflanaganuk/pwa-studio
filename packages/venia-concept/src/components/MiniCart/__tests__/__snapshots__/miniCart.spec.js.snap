// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct tree 1`] = `
<aside
  style={
    Object {
      "--minicart-height-unit": "7.19px",
    }
  }
>
  <Header
    isEditingItem={false}
  />
  <Body
    beginEditItem={[MockFunction]}
    cart={
      Object {
        "isEditingItem": false,
      }
    }
    endEditItem={[MockFunction]}
  />
  <MiniCartMask />
  <Footer
    cart={
      Object {
        "isEditingItem": false,
      }
    }
  />
</aside>
`;
